opencog/cython/opencog/backwardchainer.pyx

```diff
@@ -38,22 +40,20 @@ cdef class BackwardChainer:
                                         <cAtomSpace*> (NULL if control_as is None else control_as.atomspace),
                                         deref(focus_set.handle))
         self._as = _as
-
-    # according to PEP 442, the below simply won't work.
-    # def __del__(self):
-    # def __dealloc__(self):
-    #    del self.chainer
-    #    self._as = None
+        self._trace_as = trace_as
+        self._control_as = control_as
 
     def do_chain(self):
         return self.chainer.do_chain()
 
     def get_results(self):
         cdef cHandle res_handle = self.chainer.get_results()
         cdef Atom result = Atom.createAtom(res_handle, self._as)
+        return result
 
-        # Delete the chainer now. There does not appear to be
-        # any other way of avoiding the mem leak. See PEP 442
-        # for details.
+    def __dealloc__(self):
+        self._control_as = None
+        self._trace_as = None
+        self._as = None
         del self.chainer
```

Comment:
I now thinking that chainer should be deleted before atomspaces. Could order in destructor be important?
@vsbogd What do you think?

Reply:
As a general programming principle, deletion should always be in the opposite order of creation. Since the chainer was created last, it should be deleted first. 

In this case, the C++ code for the chainer is holding references to the atomspace. (the BIT class...) when you delete the atomspace first, you are left with a chainer pointing at deleted memory. If you are lucky, nothing bad happens. If you are unlucky, something in the chainer accidentally touches that deleted memory, and a crash results.  That's the basic reason for the generic "last created first deleted" policy, in general. 

There are two alternatives to  "last created first deleted":

1. don't store references

2. use RAII

Reply:
There is a case when Cython AtomSpace instance owns the pointer to the C++ AtomSpace instance and deletes C++ instance when Cython object is being collected. So I would propose change the order of statements in `__dealloc__` to:
```Python
def __dealloc__(self):
    del self.chainer
    def __dealloc__(self):
    ...
```

Reply:
Ok. Also it seems necessary to keep reference to child atomspace in cython atomspace wrapper class. Otherwise parent could possibly get deleted somehow before child.

Reply:
> Also it seems necessary to keep reference to parent atomspace in cython atomspace wrapper class. Otherwise parent could possibly get deleted somehow before child. 

There is no case, when Cython `AtomSpace` class owns C++ `AtomSpace` with pointer to parent.

Reply:
It is quite possible:

```
a = opencog.atomspace.AtomSpace()
b = opencog.atomspace.create_
opencog.atomspace.create_child_atomspace(  opencog.atomspace.create_value_by_type(    
b = opencog.atomspace.create_child_atomspace(  opencog.atomspace.create_value_by_type(    
 b = opencog.atomspace.create_child_atomspace(a)
del a
terminate called after throwing an instance of 'opencog::RuntimeException'
  what():  AtomTable - deleteing atomtable with subtables! (/home/noskill/projects/atomspace/opencog/atomspace/AtomTable.cc:128)
Aborted (core dumped)
```

I'll make pr with fix